src/main/java/com/sugarcrm/candybean/webservices/WS.java

```diff
@@ -5,232 +5,251 @@
+		return request(op, uri, headers, body, ContentType.APPLICATION_JSON);
 	}
-	
-	public static Map<String, Object> request(OP op, String uri, Map<String, String> headers, String body, ArrayList<HashMap<String, String>> postHeaders) throws Exception {		
-		HttpUriRequest request = null;
-	    Map<String, Object> mapParse = null;
+
+	/**
+	 * Send a DELETE, GET, POST, or PUT http request, intended for multipart data and JSON requests
+	 * @param op The type of http request
+	 * @param uri The http endpoint
+	 * @param headers Map of header key value pairs
+	 * @param body Map of Key Value pairs
+	 * @param contentType The intended content type of the body
+	 * @return Key Value pairs of the response
+	 * @throws Exception If HTTP request failed
+	 */
+	public static Map<String, Object> request(OP op, String uri, Map<String, String> headers, Map<String, Object> body, ContentType contentType) throws Exception {
 		switch (op) {
 			case DELETE:
-				request = new HttpDelete(uri);
-	            mapParse = requestIt(request, headers);
-				break;
+				return handleRequest(new HttpDelete(uri), headers);
 			case GET:
-				request = new HttpGet(uri);
-	            mapParse = requestIt(request, headers);
-				break;
+				return handleRequest(new HttpGet(uri), headers);
 			case POST:
 				HttpPost post = new HttpPost(uri);
-				
-				if (postHeaders != null) {
-					for (HashMap<String, String> h : postHeaders) {
-						for (String key : h.keySet()) {
-							post.setHeader(key, h.get(key));
+				if (body != null) {
+					if (contentType == ContentType.MULTIPART_FORM_DATA) {
+						MultipartEntityBuilder builder = MultipartEntityBuilder.create();
+						for (Map.Entry<String, Object> entry : body.entrySet()) {
+							builder.addTextBody(entry.getKey(), (String) entry.getValue());
 						}
+						post.setEntity(builder.build());
+					} else {
+						JSONObject jsonBody = new JSONObject(body);
+						StringEntity strBody = new StringEntity(jsonBody.toJSONString(), ContentType.APPLICATION_JSON);
+						post.setEntity(strBody);
 					}
 				}
-
-	            mapParse = postRequest(post, headers, body); 
-				break;
+				return handleRequest(post, headers);
 			case PUT:
 				HttpPut put = new HttpPut(uri);
 				if (body != null) {
-					put.setEntity(new StringEntity(body));
+					if (contentType == ContentType.MULTIPART_FORM_DATA) {
+						MultipartEntityBuilder builder = MultipartEntityBuilder.create();
+						for (Map.Entry<String, Object> entry : body.entrySet()) {
+							builder.addTextBody(entry.getKey(), (String) entry.getValue());
+						}
+						put.setEntity(builder.build());
+					} else {
+						JSONObject jsonBody = new JSONObject(body);
+						StringEntity strBody = new StringEntity(jsonBody.toJSONString(), ContentType.APPLICATION_JSON);
+						put.setEntity(strBody);
+					}
 				}
-				request = put;
-	            mapParse = requestIt(request, headers);
-				break;
+				return handleRequest(put, headers);
```

Comment:
The PUT and POST pretty much have the same code inside the case. We should refactor this. Put the code out and put it in a private method and you can represent `HttpPut` and `HttpPost` as `HttpEntityEnclosingRequestBase`.